### ISSUE TITLE
bug report: attack successful rate of VirtualAdversarialMethod, L2 norm version of  MomentumIterativeMethod, L2 norm of BasicIterativeMethod and L inf norm of deep fool  is quite low

### ISSUE THREAD

**Issue by sharpstill** [tensorflow/cleverhans#948]
I use VirtualAdversarialMethod to attack CIFAR-10 dataset,  but on the train dataset, the successful rate is only 0.43622. I use the default parameter setting by the way.
Can you give an VAT method example ? How to set parameter and increase the successful rate above 90%

**Comment by npapernot** [comment]
Given that this is neither a bug report or feature request, this type of question is best asked on StackOverflow so others can help provide an answer: https://stackoverflow.com/search?q=cleverhans

Would you be able to repost your question on StackOverflow and provide additional details? (e.g., what is the range of your pixel values?)

**Comment by sharpstill** [comment]
@npapernot I think the code of 
VirtualAdversarialMethod  attack has a bug inside it.

**Comment by goodfeli** [comment]
if this is actually a bug report, can you post details?

is there anything that looks wrong in the code?

are there published adversarial error rates for this algorithm that this code isn't achieving?

VAT is usually used for semi-supervised learning, not adversarial attacks, so I'm not actually sure what error rate it should get. It's likely that the numbers are lower than you're expecting because VAT isn't an aggressive attack algorithm or because you need to use different hyperparameters for this model / dataset.

**Comment by sharpstill** [comment]
@goodfeli OK, I will do more test on the algorithm on CIFAR-10 dataset, And write a report pdf file to you? Is that OK?

**Comment by npapernot** [comment]
It will likely be more useful if you can write a small script that reproduces the bug, so we can test it out on our machines as well.

**Comment by sharpstill** [comment]
@npapernot just wait for me to write it.

**Comment by sharpstill** [comment]
@goodfeli @npapernot 
I found 4 attack method in Cifar-10 untargeted attack has low successful rate, which are
1. VirtualAdversarialMethod, 2. L2 norm of MomentumIterativeMethod,
3. L2 norm of BasicIterativeMethod, 4. L infinity version of deep_fool(the original cleverhans deepfool is L2 norm, I extend it to L-inf version according to original paper,

In addition, I also test a L-infinity version of DeepFool, this code is just little modifed of original cleverhans' DeepFool to add L-inf functionality, which is described in original paper's page 5.
paper can be downloaded from https://arxiv.org/abs/1511.04599.  But it also test FAILED! SUCCESSFUL RATE is so LOW!
 implementation of L-infinity version of DeepFool, just modified little code of original L2 version cleverhans deepfool.
All of the code is wrtten in the uploaded script:
[attacker_bug_report_script.zip](https://github.com/tensorflow/cleverhans/files/2783221/attacker_bug_report_script.zip)
You can download it and just run, it will first train a model on clean CIFAR-10 data.
Then use this model to generate adversarial examples of VirtualAdversarialMethod, L2 norm of MomentumIterativeMethod, L2 norm of BasicIterativeMethod, infinity version of deep_fool. one by one.
The successful rate will be printed.

**Comment by sharpstill** [comment]
Because I am in hurry for doing an adversarial example research, if you found the bug or the problem is, please feel free to contact with me as soon as possible. My email is :  sharpstill@163.com

**Comment by sharpstill** [comment]
@goodfeli @npapernot Do you have any progress on this bug? Or do you have some feedback or problem need to let me know?

**Comment by goodfeli** [comment]
Can you explain a specific success rate that you expect for a specific model and set of success hyperparameters, why you expect that success rate (e.g. it was published before), and what success rate you're actually getting?

**Comment by sharpstill** [comment]
@goodfeli My concept of  success rate is : How many proportions of generated adversarial images can be misclassified by the classifier ? For example, the proportions  of VAT generated examples which is misclassified by the classifier (which lead the classifier to predict label other than ground truth label) is about 46%. My expectation is above 90%. and L2 norm version of BasicIterativeMethod is much lower , less than 10%. But In contrast, The success rate  of L-inf norm of  BasicIterativeMethod  is 96%.

**Comment by sharpstill** [comment]
@goodfeli all the code I have summarized to the .py file, which is uploaded in my last comment. You can download can run it, and it also supplies all the hyper parameters I used in my experiments.

**Comment by sharpstill** [comment]
for example, if you use the below configuration and attack methods, you will achieve definitely low attack successful rate on CIFAR-10.
```
ATTACKERS = {"VAT": VirtualAdversarialMethod, "MI_FGSM_L2": MomentumIterativeMethod,
             "BIM_L2": BasicIterativeMethod, "deep_fool_L_infinity":DeepFool,
             }
ATTACK_PARAM = {"VAT": {"nb_iter":20, "clip_min":-3., "clip_max":3., "eps":6.0, "xi":1e-6},
                "BIM_L2" :{"ord":2, "rand_init":False, "rand_minmax":0, "nb_iter":40,"eps_iter":0.1, "clip_min":-3., "clip_max":3.},
                "MI_FGSM_L2": {"ord": 2, "eps": 0.3, "nb_iter": 40, "decay_factor": 1.0, "eps_iter": 0.1,
                               "clip_min": -3., "clip_max": 3., },
                "deep_fool_L_infinity":{ "overshoot": 0.02, "max_iter": 40, "clip_min": -3., "clip_max": 3., "batch_size": 500, "Lp_norm": np.inf},
                }
```

**Comment by npapernot** [comment]
Could you help us by telling us what success rates you observe in your script and why you think they are low (for example, a success rate that is reported in another paper)?

**Comment by goodfeli** [comment]
I assume this is just VAT working as intended. The VAT papers don't really use it as an attack or measure its success rate AFAIK. VAT is more of a dataset augmentation technique for semi-supervised learning. I'll close this issue, but we can re-open it if anyone provides evidence that correctly implemented VAT should have a high attack success rate.

**Comment by sharpstill** [comment]
@goodfeli OK, but the other 3 L2 norm version of attack algorithm still have lower than 10%  attack successful rate:
 "BIM_L2" :{"ord":2, "rand_init":False, "rand_minmax":0, "nb_iter":40,"eps_iter":0.1, "clip_min":-3., "clip_max":3.},
 "MI_FGSM_L2": {"ord": 2, "eps": 0.3, "nb_iter": 40, "decay_factor": 1.0, "eps_iter": 0.1,
                               "clip_min": -3., "clip_max": 3., },
"deep_fool_L_infinity":{ "overshoot": 0.02, "max_iter": 40, "clip_min": -3., "clip_max": 3., "batch_size": 500, "Lp_norm": np.inf}

Can you reopen this question, because I think there still be bug inside BIM and momentum iterative GSM at least.

**Comment by sharpstill** [comment]
@npapernot in VAT attack, I observe the  attack successful rate is 46%. My expectation is above 90%. And in L2 norm of BIM and L2 norm of momentum iterative FGSM (MI-FGSM),   both attacks achieve successful rate less than 10%.   But the successful rate of  L-inf norm of BIM and MI-FGSM is 98%.

**Comment by sharpstill** [comment]
@npapernot BTW, I use CIFAR_10 to conduct my experiment, please download my script file to run .
[attacker_bug_report_script.zip](https://github.com/tensorflow/cleverhans/files/2816005/attacker_bug_report_script.zip)
